# Projeto 2 - Classificador Automático de Sentimento

Você foi contratado por uma empresa parar analisar como os clientes estão reagindo a um determinado produto no Twitter. A empresa deseja que você crie um programa que irá analisar as mensagens disponíveis e classificará como "relevante" ou "irrelevante". Com isso ela deseja que mensagens negativas, que denigrem o nome do produto, ou que mereçam destaque, disparem um foco de atenção da área de marketing.<br /><br />
Como aluno de Ciência dos Dados, você lembrou do Teorema de Bayes, mais especificamente do Classificador Naive-Bayes, que é largamente utilizado em filtros anti-spam de e-mails. O classificador permite calcular qual a probabilidade de uma mensagem ser relevante dadas as palavras em seu conteúdo.<br /><br />
Para realizar o MVP (*minimum viable product*) do projeto, você precisa implementar uma versão do classificador que "aprende" o que é relevante com uma base de treinamento e compara a performance dos resultados com uma base de testes.<br /><br />
Após validado, o seu protótipo poderá também capturar e classificar automaticamente as mensagens da plataforma.

## Informações do Projeto

Prazo: 13/Set até às 23:59.<br />
Grupo: 1 ou 2 pessoas.<br /><br />
Entregáveis via GitHub: 
* Arquivo notebook com o código do classificador, seguindo as orientações abaixo.
* Arquivo Excel com as bases de treinamento e teste totalmente classificado.

**NÃO disponibilizar o arquivo com os *access keys/tokens* do Twitter.**


### Check 3: 

Até o dia 06 de Setembro às 23:59, o notebook e o xlsx devem estar no Github com as seguintes evidências: 
    * Conta no twitter criada.
    * Produto escolhido.
    * Arquivo Excel contendo a base de treinamento e teste já classificado.
    

Sugestão de leitura:<br />
http://docs.tweepy.org/en/v3.5.0/index.html<br />
https://monkeylearn.com/blog/practical-explanation-naive-bayes-classifier/

___

## Preparando o ambiente

Instalando a biblioteca *tweepy* para realizar a conexão com o Twitter:

In [ ]:
%%capture

#Instalando o tweepy
!pip install tweepy

Importando as Bibliotecas que serão utilizadas. Esteja livre para adicionar outras.

In [1]:
import tweepy
import math
import os.path
import pandas as pd
import json
from random import shuffle

___
## Autenticando no  Twitter

Para realizar a captura dos dados é necessário ter uma conta cadastrada no twitter:

* Conta: ***[Preencha aqui o id da sua conta. Ex: @fulano ]***


1. Caso ainda não tenha uma: https://twitter.com/signup
1. Depois é necessário registrar um app para usar a biblioteca: https://apps.twitter.com/
1. Dentro do registro do App, na aba Keys and Access Tokens, anotar os seguintes campos:
    1. Consumer Key (API Key)
    1. Consumer Secret (API Secret)
1. Mais abaixo, gere um Token e anote também:
    1. Access Token
    1. Access Token Secret
    
1. Preencha os valores no arquivo "auth.pass"

**ATENÇÃO**: Nunca divulgue os dados desse arquivo online (GitHub, etc). Ele contém as chaves necessárias para realizar as operações no twitter de forma automática e portanto é equivalente a ser "hackeado". De posse desses dados, pessoas mal intencionadas podem fazer todas as operações manuais (tweetar, seguir, bloquear/desbloquear, listar os seguidores, etc). Para efeito do projeto, esse arquivo não precisa ser entregue!!!

In [ ]:
#Dados de autenticação do twitter:

#Coloque aqui o identificador da conta no twitter: @fulano

#leitura do arquivo no formato JSON
with open('auth.pass') as fp:    
    data = json.load(fp)

#Configurando a biblioteca. Não modificar
auth = tweepy.OAuthHandler(data['consumer_key'], data['consumer_secret'])
auth.set_access_token(data['access_token'], data['access_token_secret'])

___
## Coletando Dados

Agora vamos coletar os dados. Tenha em mente que dependendo do produto escolhido, não haverá uma quantidade significativa de mensagens, ou ainda poder haver muitos retweets.<br /><br /> 
Configurando:

In [ ]:
#Produto escolhido:
produto = 'LATAM'

#Quantidade mínima de mensagens capturadas:
n = 500
#Quantidade mínima de mensagens para a base de treinamento:
t = 300

#Filtro de língua, escolha uma na tabela ISO 639-1.
lang = 'pt'

Capturando os dados do twitter:

In [ ]:
#Cria um objeto para a captura
api = tweepy.API(auth)

#Inicia a captura, para mais detalhes: ver a documentação do tweepy
i = 1
msgs = []
for msg in tweepy.Cursor(api.search, q=produto, lang=lang).items():    
    msgs.append(msg.text.lower())
    i += 1
    if i > n:
        break

#Embaralhando as mensagens para reduzir um possível viés
shuffle(msgs)

Salvando os dados em uma planilha Excel:

In [ ]:
#Verifica se o arquivo não existe para não substituir um conjunto pronto
if not os.path.isfile('./{0}.xlsx'.format(produto)):
    
    #Abre o arquivo para escrita
    writer = pd.ExcelWriter('{0}.xlsx'.format(produto))

    #divide o conjunto de mensagens em duas planilhas
    dft = pd.DataFrame({'Treinamento' : pd.Series(msgs[:t])})
    dft.to_excel(excel_writer = writer, sheet_name = 'Treinamento', index = False)

    dfc = pd.DataFrame({'Teste' : pd.Series(msgs[t:])})
    dfc.to_excel(excel_writer = writer, sheet_name = 'Teste', index = False)

    #fecha o arquivo
    writer.save()

___
## Classificando as Mensagens

Agora você deve abrir o arquivo Excel com as mensagens capturadas e classificar na Coluna B se a mensagem é relevante ou não.<br /> 
Não se esqueça de colocar um nome para a coluna na célula **B1**.<br /><br />
Fazer o mesmo na planilha de Controle.

___
## Montando o Classificador Naive-Bayes

Com a base de treinamento montada, comece a desenvolver o classificador. Escreva o seu código abaixo:

Opcionalmente: 
* Limpar as mensagens removendo os caracteres: enter, :, ", ', (, ), etc. Não remover emojis.<br />
* Corrigir separação de espaços entre palavras e/ou emojis.
* Propor outras limpezas/transformações que não afetem a qualidade da informação.



In [2]:
# ABRINDO O EXCEL DE TREINAMENTO PARA O CLASSIFICADOR

excel_latam = pd.read_excel('Treinamento.xlsx')

In [3]:
relevantes = excel_latam[excel_latam.Relev==1] #Criando um DataFrame só com os tweets relevantes
irrelevantes = excel_latam[excel_latam.Relev==0] #Criando um DataFrame só com os tweets irrelevantes

In [4]:
# FILTRANDO TODAS PALAVRAS NOS TWEETS RELEVANTES COM E SEM REPETIÇÃO

lista_relevantes = [] #lista de todas palavras nos tweets relevantes
nao_repetidas_relevantes = [] #lista de palavras nos tweets relevantes sem repetição
remover=["!", ";", "@", ".", ",", "'","?", "#"] #lista de caracteres indesejados para nosso modelo remover

for i in relevantes.Tweets: #código qu adiciona todas palavras em uma lista das palavras relevantes
    a=i.split()
    for j in a:
        r=list(j)
        for u in r:
            if u in remover:
                r.remove(u)
        e=''.join(r)
        lista_relevantes.append(e)
        
        
for t in lista_relevantes: #código que adiciona todas palavras encontradas relevantes em uma lista, porém SEM repetição
    if t not in nao_repetidas_relevantes:
        nao_repetidas_relevantes.append(t)
        
numero_nao_repetidas_relevantes=len(nao_repetidas_relevantes) #definindo o espaço amostral das palavras relevantes

In [5]:
# FILTRANDO TODAS PALAVRAS NOS TWEETS IRRELEVANTES COM E SEM REPETIÇÃO

lista_irrelevantes = [] #lista de todas palavras nos tweets irrelevantes
nao_repetidas_irrelevantes = [] #lista de palavras nos tweets irrelevantes sem repetição

for i in irrelevantes.Tweets: #código que adiciona todas palavras em uma lista das palavras irrelevantes
    a=i.split()
    for j in a:
        r=list(j)
        for u in r:
            if u in remover:
                r.remove(u)
        e=''.join(r)
        lista_irrelevantes.append(e)
                
for t in lista_irrelevantes:  #código que adiciona todas palavras irrelevantes encontradas em uma lista, porém SEM repetição
    if t not in nao_repetidas_irrelevantes:
        nao_repetidas_irrelevantes.append(t)
        
numero_nao_repetidas_irrelevantes=len(nao_repetidas_irrelevantes) #definindo o espaço amostral das palavras irrelevantes

In [6]:
# CONTAGEM DE PALAVRAS NO TOTAL PARA CASOS RELEVANTES E IRRELEVANTES

palavras_total = [] #lista de todas palavras nos tweets
palavras_nao_repetidas = [] #lista de palavras nos tweets sem repetição

for i in excel_latam.Tweets: #código que adiciona todas palavras em uma lista das palavras
    a=i.split()
    for j in a:
        r=list(j)
        for u in r:
            if u in remover:
                r.remove(u)
        e=''.join(r)
        palavras_total.append(e)
                
for t in palavras_total: #código que adiciona todas palavras encontradas em uma lista, porém SEM repetição
    if t not in palavras_nao_repetidas:
        palavras_nao_repetidas.append(t)
        
numero_nao_repetidas=len(palavras_nao_repetidas)

In [7]:
# DEFININDO A FUNÇÃO QUE CALCULA PROBABILIDADE, RECEBENDO UMA FRASE (TWEET) COMO PARÂMETRO
# A FUNÇÃO RETORNA "RELEVANTE" SE A PROBABILIDADE DO TWEET SER RELEVANTE FOR MAIOR DO QUE A PROBABILIDADE SE FOR IRRELEVANTE
# A FUNÇÃO RETORNA "IRRELEVANTE" SE A PROBABILIDADE DO TWEET SER IRRELEVANTE FOR MAIOR DO QUE A PROBABILIDADE SE FOR RELEVANTE

def prob(frase):
    lista_pr = []
    lista_pi = []
    probrel=1
    probirrel=1
    somar=0
    somai=0
    split=frase.split()
    for i in split:
        if i in lista_relevantes:
            somar+=1
        pr=(1+somar)/(numero_nao_repetidas+numero_nao_repetidas_relevantes)
        lista_pr.append(pr)
    for y in range(len(lista_pr)):
        probrel=probrel*lista_pr[y]
    
    for f in split:
        if f in lista_irrelevantes:
            somai+=1
        pi=(1+somai)/(numero_nao_repetidas+numero_nao_repetidas_irrelevantes)
        lista_pi.append(pi)
    for w in range(len(lista_pi)):
        probirrel=probirrel*lista_pi[w]
        
    if probrel > probirrel:
        return "Relevante"
    elif probrel < probirrel:
        return "Irrelevante"

___
## Verificando a performance

Agora você deve testar o seu Classificador com a base de Testes.<br /><br /> 

Você deve extrair as seguintes medidas:
* Porcentagem de positivos falsos (marcados como relevante mas não são relevantes)
* Porcentagem de positivos verdadeiros (marcado como relevante e são relevantes)
* Porcentagem de negativos verdadeiros (marcado como não relevante e não são relevantes)
* Porcentagem de negativos falsos (marcado como não relevante e são relevantes)

Opcionalmente:
* Criar categorias intermediárias de relevância baseado na diferença de probabilidades. Exemplo: muito relevante, relevante, neutro, irrelevante e muito irrelevante.

In [8]:
# ABRINDO O EXCEL DE TESTE PARA TESTAR O CLASSIFICADOR
excel_teste = pd.read_excel('Teste.xlsx')

# CRIANDO UM DATAFRAME SÓ COM OS TWEETS E OUTRO DATAFRAME SÓ COM A RELEVÂNCIA

teste_tweets=pd.DataFrame(excel_teste.loc[:,["Teste"]])
teste_relev=pd.DataFrame(excel_teste.loc[:,["Relev"]])

In [11]:
# CRIANDO A FUNÇÃO QUE PEGA TODOS TWEETS E PASSA UM A UM EM LOOP NO CLASSIFICADOR, ARMAZENA SE O CLASSIFICADOR ACERTOU OU ERROU
# E IDENTIFICA O TIPO DE ACERTO E O TIPO DE ERRO (POSITIVO FALSO, NEGATIVO VERDADEIRO, ETC...) E CALCULA A PORCENTAGEM DE CADA
# UM DESSES PARÂMETROS, ALÉM DE RETORNAR A SOMA DOS ACERTOS E DOS ERROS DO CLASSIFICADOR

def acertos():
    somapv = 0
    somapf=0
    somanv=0
    somanf=0
    relev=[]
    lista_teste_tweets=[]
    for u in teste_relev.Relev:
        relev.append(u)
    for j in teste_tweets.Teste:
        lista_teste_tweets.append(prob(j))
    for y in range(len(relev)):
        if relev[y]==1:
            if lista_teste_tweets[y]=="Relevante":
                somapv+=1
            if lista_teste_tweets[y]=="Irrelevante":
                somanf+=1
        if relev[y]==0:
            if lista_teste_tweets[y]=="Relevante":
                somapf+=1
            if lista_teste_tweets[y]=="Irrelevante":
                somanv+=1
    probpv=((somapv/200)*100)
    probpf=((somapf/200)*100)
    probnv=((somanv/200)*100)
    probnf=((somanf/200)*100)
    proba=probpv+probnv
    probe=probpf+probnf
    
    return probpv, probpf, probnv, probnf, proba, probe

In [13]:
# RODANDO A FUNÇÃO ACERTOS

acertos()

(16.0, 42.5, 41.5, 0.0, 57.5, 42.5)

### Porcentagem de positivos verdadeiros: 16%
### Porcentagem de positivos falsos: 42,5%
### Porcentagem de negativos verdadeiros: 41,5%
### Porcentagem de negativos falsos: 0%


### Total de acertos: 57,5%
### Total de erros: 42,5%

___
## Concluindo

Escreva aqui a sua conclusão.<br /> 
Faça um comparativo qualitativo sobre as medidas obtidas.<br />
Explique como são tratadas as mensagens com dupla negação e sarcasmo.<br />
Proponha um plano de expansão. Por que eles devem continuar financiando o seu projeto?<br />

Opcionalmente: 
* Discorrer por que não posso alimentar minha base de Treinamento automaticamente usando o próprio classificador, aplicado a novos tweets.
* Propor diferentes cenários de uso para o classificador Naive-Bayes. Cenários sem intersecção com este projeto.
* Sugerir e explicar melhorias reais no classificador com indicações concretas de como implementar (não é preciso codificar, mas indicar como fazer e material de pesquisa sobre o assunto).
